mirgecom/flux.py

```diff
@@ -3,7 +3,8 @@
 Numerical Flux Routines
 ^^^^^^^^^^^^^^^^^^^^^^^
 
-.. autofunction:: lfr_flux
+.. autofunction:: gradient_flux_central
+.. autofunction:: divergence_flux_lfr
```

Comment:
`divergence_flux_lfr` sounds strange. Technically none of these flux functions should care (in principle) whether you're computing fluxes for the primary or auxiliary equations. I know why you'd want to distinguish between the primary and auxiliary equations in DG, but I feel like it may be better to just leave `flux_lfr` unchanged for now. Thoughts?

Reply:
Not sure I follow... Note that the current LFR implementation here encodes a divergence with the `@ normal`; it's not quite the same as the book.

Reply:
I share your perspective on this @thomasgibson, I think.  My take is that regardless of what the operator or set of equations is, one will be either integrating a vector or a scalar at the element boundary (i.e. either `a*n_hat` or `v.dot.nhat`), and it made sense to me to name them accordingly.  The different element flux routines seem to just come down to how the `a` or `v` get reconstructed before scaling, or dotting (respectively) `n_hat`).

FWIW, I am just fine leaving this how it is. It doesn't break anything, it isn't wrong,  and we can talk at length about how to name them.

Reply:
> Not sure I follow... Note that the current LFR implementation here encodes a divergence with the `@ normal`; it's not quite the same as the book.

In my H+W, in the Beyond One Dimension chapter, where they implement a 2D Euler solver, they lay out the Lax-Friedrichs as this:

`{{Flux}}.dot.n_hat + lam[[CV]]/2`

This is also how it is implemented in their example code in both the 1D and the 2D example, so I duplicated.

Reply:
> In my H+W, in the Beyond One Dimension chapter, where they implement a 2D Euler solver, they lay out the Lax-Friedrichs as this:
> 
> `{{Flux}}.dot.n_hat + lam[[CV]]/2`
> 
> This is also how it is implemented in their example code in both the 1D and the 2D example, so I duplicated.

Sorry, I should have been more specific; I meant the general form that the book describes (the one I screenshotted in the other discussion thread). The version used for Euler is just that applied to the divergence operator. I'm guessing @thomasgibson is thinking about the more general version here (?). If we had a function that computed the general form, it would make sense to call it `lfr_flux`, but since it's currently specific to divergence I think `divergence_flux_lfr` is appropriate.

Reply:
I am not sure.  I think the one you screenshotted was a 1D version? It is extended to multiple dimensions as you read on in the same chapter - wrapping up in section 2.3 with eqns 2.16,2.17 and the following definition of the Lax-Friedrichs (the one we currently implement).

Reply:
> Edit: I think I see now, Matt. You mean I'm returning (f*.dot.n, as appropriate for the divergence operator) and not just f*?

Yep, exactly.

Reply:
#494 🤔

Reply:
> #494 🤔

Neat. Yeah, I like this. It would be cool to eventually generalize this, adding a `flux_central`, etc. Then we could have a single flux function for each operator and pass these guys in as parameters.

Reply:
Trouble is that it does not work as intended.  I have not been able to piece together what is going wrong yet.  I think this part at least could be safely pushed off to a dedicated flux-refactoring effort.  Whadda you think?

Reply:
> Trouble is that it does not work as intended. I have not been able to piece together what is going wrong yet. I think this part at least could be safely pushed off to a dedicated flux-refactoring effort. Whadda you think?

https://github.com/illinois-ceesd/mirgecom/pull/494#discussion_r694926647

Reply:
> > Trouble is that it does not work as intended. I have not been able to piece together what is going wrong yet. I think this part at least could be safely pushed off to a dedicated flux-refactoring effort. Whadda you think?
> 
> [#494 (comment)](https://github.com/illinois-ceesd/mirgecom/pull/494#discussion_r694926647)

Awesome!  You fixed it! This exact issue has been plaguing me for some time.  It is why the LFR flux routine is how it ~is~ was.  Otherwise, I would have applied the .dot.normal up inside `inviscid_facial_flux` - but I could not make that work at the time because I couldn't figure out how to form f^* in a way that worked.

Reply:
@majosm The tests no longer work right after the LFR change. I am not sure what is wrong here, PyOpenCL claims to run out of memory?!

Reply:
Yeah. It looks like it's only failing for `test_multilump_rhs` for order 4 in 3D (which has a largest mesh size of 16 elements per axis). My guess is that it's either running out of memory due to a bunch of temporary arrays floating around or the memory is just getting too fragmented. I'm trying a CI run with a memory pool allocator to see if fragmentation is the issue: #496